src/validator.js

```diff
@@ -75,6 +83,68 @@ export default class Validator {
     return output;
   }
 
+  scan(_Xpi=Xpi) {
+    return new Promise((resolve, reject) => {
+      this.checkFileExists(this.packagePath)
+        .then(() => {
+          this.xpi = new _Xpi(this.packagePath);
+          return this.xpi.getMetaData();
+        }).then(() => {
+          return this.xpi.getJSFiles();
+        })
+        .then((jsFiles) => {
+          return this.scanJSFiles(jsFiles);
+        })
+        .then(() => {
+          this.print();
+          resolve();
+        })
+        .catch((err) => {
+          if (err instanceof exceptions.DuplicateZipEntryError) {
+            this.collector.addError(messages.DUPLICATE_XPI_ENTRY);
+            this.print();
+          } else {
+            this.handleError(err);
+          }
+
+          reject(err);
+        });
+    });
+  }
+
+  scanJSFiles(jsFiles) {
+    return new Promise((resolve, reject) => {
+      // Resolve once every file in the XPI has been checked.
+      var jsFilesPromises = [];
+
+      jsFiles.forEach((filename) => {
+        jsFilesPromises.push(this.scanJSFile(filename));
+      });
+
+      return Promise.all(jsFilesPromises).then(() => {
```

Comment:
It’s typically not how I’ve written stuff but I can totally get on board, and I’m for consistency > all else :-)

I’ll see if I can wire up a JSCS rule to prevent me from doing that without getting yelled at by the linter :)